### DISCUSSION TITLE
Add a way to know if a WebP file is lossless with vipsheaders

### DISCUSSION THREAD

**Discussion by jfcalvo** [libvips/libvips#2406] [Ideas]
I have not found any way to check if a specific WebP file is lossless using `vipsheaders` command.

If there is not other way to check this I think it could be a good idea to add a header like `lossless: 1` or similar to indicate this.

Thanks.

**Comment by jcupitt** [comment]
I agree, this is a good idea.

**Comment by lovell** [comment]
WebP supports a mixture of "formats" within the same image file, so a rather contrived example might have frame 1 with lossy RGB and lossless alpha, frame 2 with all lossless, frame 3 with lossless RGB and lossy alpha.

The libwebp API provides "lossy", "lossless" or "mixed" via `WebPBitstreamFeatures` - see https://github.com/webmproject/libwebp/blob/36a6eea3bc2871c7f0166193a6fa42241af4a464/src/webp/decode.h#L417

I guess libvips could expose this as an array of values, but for the example I don't think we'd be able to differentiate between frames 1 ("mixed") and 3 (also "mixed").

**Reply by jcupitt** [reply]
Oh, true. A very common case is lossy RGB, but lossless alpha.

You're right, I think this would be complex to support fully. @jfcalvo, would you mind explaining the use case you have in mind?

**Reply by jfcalvo** [reply]
I didn't know that something like mixed frames (lossy and lossless) was something possible with WebP. Maybe exposing the `format` suggested by @lovell as an array is not a bad idea (but not sure about the complexity of this).

I'm writing tests for an application that I'm building so I was trying to find an easy way to know if a WebP file has been saved with `lossless` argument to `true` (in my case animated and non animated files).

**Reply by jcupitt** [reply]
You can spot animations by looking for the `n-pages` property, if that helps.

Maybe you could just add `lossless` to the filename? Assuming you're the person saving these files.

**Reply by jfcalvo** [reply]
Not sure if changing the filename is useful. What I'm trying to check is that we are really encoding the WebP as lossless so it can fake the results.

I'm experimenting with `webpinfo` command and is showing me the following (for a lossless encoded WebP with 1 frame and no alpha):
```
$ webpinfo example.webp
File: example.webp
RIFF HEADER:
  File size: 303464
Chunk VP8X at offset     12, length     18
  ICCP: 1
  Alpha: 0
  EXIF: 1
  XMP: 1
  Animation: 0
  Canvas size 500 x 334
Chunk ICCP at offset     30, length   3152
Chunk VP8L at offset   3182, length 271962
  Width: 500
  Height: 334
  Alpha: 0
  Animation: 0
  Format: Lossless (2)
Chunk EXIF at offset 275144, length  20040
Chunk XMP  at offset 295184, length   8280
No error detected.
```

It's printing `Format: Lossless (2)`. Not sure how difficult could be to replicate something similar with `vipsheader`.

Please feel free to close the discussion if this is too difficult to implement or the benefits are not clear.

Thank you very much for the suggestions as always.

**Comment by tats-u** [comment]
I want this metadata to resize and re-compress WebP.
Lossy re-compression of lossless WebP spoils colors of thin lines and texts.

Where should we add `vips_image_set_*` in `read_header` in `webp2vips.c`?
What is the best for the lossless flag in the meta data? `lossless` will make us implement for other image types than WebP.
Also, there seems to be no `vips_image_set_boolean`.

**Reply by jcupitt** [reply]
Hi @tats-u,

Perhaps just after the code that sets XMP/ICC/exif metadata?

https://github.com/libvips/libvips/blob/master/libvips/foreign/webp2vips.c#L567

But it might be a little difficult to detect if a webp image is lossless. You need to check the encoding of every image object in every frame, and that will probably be quite slow.

> Also, there seems to be no `vips_image_set_boolean()`.

We use an int with a value of 0 or 1, eg.:

https://github.com/libvips/libvips/blob/master/libvips/foreign/jpeg2vips.c#L645-L648

**Reply by jcupitt** [reply]
Even a single frame can be mixed lossy and lossless -- it's very common to use lossy RGB but lossless alpha, for example, to avoid banding in dark areas.

**Reply by tats-u** [reply]
@jcupitt

Sorry for very late reply. I may try when I have a time.

> You need to check the encoding of every image object in every frame, and that will probably be quite slow.

How about limiting to the first frame? It should minimize the bad influence on the performance.
If users want to check if the image is lossless strictly, they should do it by an independent library by themselves.

**Reply by tats-u** [reply]
>  lossy RGB but lossless alpha

If it's going to be converted to AVIF, its chroma should be 4:2:0, not 4:4:4. It's closer to JPEG than PNG.

How should we treat lossless RGB with lossy alpha? Which is it closer, JPEG or PNG?

> limiting to the first frame

This is intended for animation WebP.
